### ISSUE TITLE
sql: panic in fuzzystrmatch in soundex builtin

### ISSUE THREAD

**Issue by ecerichter** [cockroachdb/cockroach#84016]
I'm trying to create a "soundex" query like

```
select *
   from logradouro
  where soundex(logradouro.nome) = soundex('Agronomia')
```

but I always get:

```
I220707 16:36:59.635917 1 util/log/flags.go:202  [-] 1  stderr capture started
panic: runtime error: index out of range [4] with length 4

goroutine 32350 [running]:
github.com/cockroachdb/cockroach/pkg/util/fuzzystrmatch.soundex({0xc005e815e8, 0xc005bd0d18})
        github.com/cockroachdb/cockroach/pkg/util/fuzzystrmatch/soundex.go:59 +0x366
github.com/cockroachdb/cockroach/pkg/util/fuzzystrmatch.Soundex({0xc005e815e8, 0xc005d2c660})
        github.com/cockroachdb/cockroach/pkg/util/fuzzystrmatch/soundex.go:77 +0x25
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.glob..func211(0xc005d12a00, {0xc0009eabc0, 0x0, 0x0})
        github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:3728 +0x4a
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc005d12a00, 0xc005a0ef00)
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:4377 +0x86
github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfiller).BuildIndexEntriesChunk.func1({0xc0009ea7e0, 0x1, 0x0})
        github.com/cockroachdb/cockroach/pkg/sql/backfill/backfill.go:840 +0xee
github.com/cockroachdb/cockroach/pkg/sql/backfill.(*IndexBackfiller).BuildIndexEntriesChunk(0xc000a90c80, {0x631e378, 0xc00b00e060}, 0x203001, {0x644a8d8, 0xc0057ef400}, {{0xc005e80d68, 0x4, 0x8}, {0xc005e80d70, ...}}, ...)
        github.com/cockroachdb/cockroach/pkg/sql/backfill/backfill.go:871 +0xbe4
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).buildIndexEntryBatch.func1({0x631e378, 0xc00b00e060}, 0xc00b00e000)
        github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:435 +0xf2
github.com/cockroachdb/cockroach/pkg/kv.runTxn.func1({0x631e378, 0xc00b00e060}, 0xc002ed9678)
        github.com/cockroachdb/cockroach/pkg/kv/db.go:949 +0x27
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc005899ce0, {0x631e378, 0xc00b00e060}, 0xc005bd1980)
        github.com/cockroachdb/cockroach/pkg/kv/txn.go:993 +0xae
github.com/cockroachdb/cockroach/pkg/kv.runTxn({0x631e378, 0xc00b00e060}, 0xc0005dcd80, 0x9121d9)
        github.com/cockroachdb/cockroach/pkg/kv/db.go:948 +0x5a
github.com/cockroachdb/cockroach/pkg/kv.(*DB).TxnWithAdmissionControl(0x631e378, {0x631e378, 0xc00b00e060}, 0xa, 0x0, 0xc00b10db00)
        github.com/cockroachdb/cockroach/pkg/kv/db.go:911 +0x89
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(...)
        github.com/cockroachdb/cockroach/pkg/kv/db.go:890
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).buildIndexEntryBatch(0xc000a90c80, {0x631e378, 0xc00b00e000}, {{0xc005e80d68, 0x4, 0x8}, {0xc005e80d70, 0x4, 0x8}}, {0x16ff9b07daf2b8ae, ...})
        github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:428 +0x2fc
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).constructIndexEntries(0xc000a90c80, {0x631e378, 0xc00b00e000}, 0xc004c74d20)
        github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:134 +0x305
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*indexBackfiller).runBackfill.func1({0x631e2d0, 0xc0062ce600})
        github.com/cockroachdb/cockroach/pkg/sql/rowexec/indexbackfiller.go:319 +0xcb
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:169 +0x25
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
        golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:54 +0x92
```

Have I faced a bug in soundex functionality?

Thanks for your help,

Edson

Jira issue: CRDB-17409

**Comment by knz** [comment]
cc @rafiss for triage

**Comment by rafiss** [comment]
I'll close this, since I think the linked PR should fix this. Please re-open if you still see this behavior when using v21.2.13 (already released) or v22.1.3 (should be available in a few days).
